velox/docs/develop/window.rst

```diff
@@ -0,0 +1,202 @@
+  have both frame ends before the first partition row.
+
+* Both frameStart and frameEnds fall after the partition end row.
+
+  E.g. in frame *ROWS BETWEEN 2 FOLLOWING and 5 FOLLOWING* the last 2 rows
+  have both frame ends beyond the last partition row.
+
+* frameStart > frameEnds row (as frame range is defined from frameStart to
+  frameEnds).
+
+  E.g. In frame *ROWS BETWEEN UNBOUNDED PRECEDING AND 2 PRECEDING* the intent
+  is to compute aggregation from the partition start row to 2 rows prior
+  the current one. However, for the first 2 rows the frameStart
+  (frame index 0 for unbounded preceding) is ahead of 2 preceding
+  (indices -2 and -1).
+
+* For frames like *ROWS BETWEEN 2 PRECEDING AND 5 PRECEDING* or
+  *ROWS BETWEEN 5 FOLLOWING AND 2 FOLLOWING*, frameStart > frameEnds for
+  all rows. So all frames are empty.
+
+**Partial frames**
+
+Like illustrated in the examples above, rows could have partial window frames.
+A partial frame occurs when:
+
+- frameStart < frameEnd (so the empty frame condition is not violated)
+- One frame end is within partition bounds and the other end outside of it.
+  In these cases either frameStarts is before partitionStart while
+  frameEnds within the partition, or frameStarts is in the partition but
+  frameEnds beyond it.
+For such frames the violating boundary is clamped to the partitionStart or
+partitionEnd.
+
+Partial frames usually follow empty frames in a sliding window.
+
+E.g. In frame *ROWS BETWEEN 5 PRECEDING AND 2 PRECEDING*, the first 2 rows have both frameEnds
+before partition start, so they are empty. But from 3rd - 5th row, the 5 preceding
+bound (frameStarts) is outside the partition, but 2 preceding (frameEnds) is within the
+partition. So for these 3 rows the frameStart is clamped to partitionStart.
+
+Similarly for frame *ROWS BETWEEN 2 FOLLOWING AND 5 FOLLOWING*, the 3-5th rows
+before partition end have the frameStart end within the partition,
+but frameEnds beyond. So they are partial frames.
+The last 2 rows have both ends outside the partition and are empty frames..
+
+Empty and partial window frames can be visualized as below
+
+.. image:: images/empty_frames.png
+    :width: 600
+    :align: center
+
+Frames with constant frame bounds (like 2 preceding) have strict sliding behavior.
+So the empty frames, partial frames and valid frames cluster together and follow
+(or precede) each other. Adhoc frames that use column values for bounds can have
+empty, partial or valid frames at any points in the partition rows.
+
+**Handling empty frames in window functions**
+
+As mentioned before, only value and aggregate window functions use frames in
+their evaluation. Both Window functions return null values for empty frames.
+
+The most naive approach to handle empty frames is to check in the window
+function logic if the frame indices are an empty frame (based on the
+conditions previously described) and return the null output.
+This is very inefficient.
```

Comment:
Wondering why is this inefficient.

Reply:
If there are multiple window functions in the same operator this logic would be evaluated multiple times. Also the code is repeated in every function. Have clarified in the doc.